### ISSUE TITLE
Python: provide inclusion/exclusion of language runtimes

### ISSUE THREAD

**Issue by jthomasmock** [posit-dev/positron#3574]
While we have robust discovery of language runtimes, it may be nice to exclude specific versions of Python, such as the system python that shouldn't be used. I've also seen instances of old Conda/homebrew installs of Python people still have around even though they want to use a tool like `pyenv`. While I generally think an exclusion is the right approach, it also might help to be able to provide a distinct inclusion list to limit the time spent searching/discovery although I think we're pretty efficient there.

One example of a report below with VS Code on Workbench:

>  VS Code is quite good in detecting all sorts of python versions on a given system which can be confusing for some users. I haven’t found a way to only offer the “desired” python versions (and for example hide the system python version that may be way too old). What we did at <COMPANY> was at least to set `python.defaultInterpreterPath` - this setting typically lives in `~/.local/share/code-server/Machine/settings.json` and hence cannot be set in `vscode-user-settings.json` which seems to only yield into `~/.local/share/code-server/User/settings.json` only. We implemented this by adding this file to each existing user’s home-directory and put the contents of `~/.local/share/code-server/Machine/settings.json`  into `/etc/skel` so it is being picked up for any new user.

---

## Implementation Draft Summary

This is the Python equivalent to https://github.com/posit-dev/positron/issues/6205. These options will interact with #6203.

New options will be added to include/exclude Python interpreters, which the user can configure via the Settings UI or via the user `settings.json` directly. As such, these settings will be applicable at the workspace and/or user scopes, and can be preset by a workbench admin via `positron-user-settings.json`.

### add `python.interpreters.include` settings option
- List of strings, where each string is a path to a specific interpreter or a directory to search for interpreters
- Draft description: List of additional interpreters to include in the set of Python interpreters we discover. You can specify a directory path that we should search for interpreters.
- Default: null

### add `python.interpreters.exclude` settings option
- List of strings, where each string is a path to a specific interpreter or a directory to NOT search for interpreters
- Draft description: List of interpreters to exclude from the available Python interpreters shown in Positron. You can specify the file path of a specific interpreter, or a directory path that we should not search for interpreters.
- Default: null (possibly include homebrew python/other system pythons as default excludes?)

### Mockup of Settings UI

![Image](https://github.com/user-attachments/assets/02c15a24-324d-4350-b746-f8705a4d2601)

### What if a user specifies invalid paths, the interpreters or directories don’t exist, or something else is wrong with the user’s input in settings?
- If any of these are invalid for any reason (i.e. broken path, does not exist), we will indicate as such in logs and/or via Notification toasts at the time of discovery, as there isn’t very robust input validation in the Settings UI.
- May need some additional UI/notifications to cover the case where this causes discovery to find nothing. “Could not find any interpreters included in the setting python.interpreters.override. Would you like Positron to try again, without the setting?”
    - We have a Positron core toast “no interpreters found”, but maybe not for Python or R yet

### What if there is overlap between the paths specified in includes/excludes?
- excludes will take precedence over includes

---

## Implemented in
- https://github.com/posit-dev/positron/pull/6398

**Comment by juliasilge** [comment]
Some thoughts on default interpreters and discovery: https://github.com/posit-dev/positron/issues/5664#issuecomment-2528464379

**Comment by ntluong95** [comment]
I have been using Positron for about 6 months now and I found Positron often take longer time than VSCode. I know that there are many bundled extensions that may slow the process a bit, but the main problem is usually the R/Python interpreter usually take longer time to be ready. I have a proposal, which can be apply for me but not the other as following:
- For a fresh, new user, the first time start it should scan all available interpreters in the computer and ask them to set the default interpreter. We can have a setting at IDE/User level, so the path to default R/Python interpreter is provided. This is a reliable way for the Console to always start without discovering all existing interpreter in the computer
- For a fresh, new project, it just loads the global R interpreter and/or Python interpreter. We can have a setting that if when starting a new project, do they want to load R/Python or both global interpreters. If they are R user or the vice versa. they may not need to load them all together. This approach is similar to Rstudio, the Python interpreter only starts once you click on the button
- When opening an old project, it should load the previous interpreter. If the user wants to change the interpreter, the discovering process will be trigger if they press the button on top-right corner. This is because if you have worked in a project before, you have defined your preferred interpreter, and you will not change it in the next time. We can also have a setting at Workspace level, to either load one of them or both for the project.

So basically, I propose 2 settings:

- Set the default interpreter
- How user want to load R and/or Python interpreter for a new/old project

Those setting will be helpful in the future if you support more languages. User should have the ability to turn-off things they do not use often :)

**Comment by juliasilge** [comment]
As pointed out in #5945, this plays out in an annoying way on Posit Workbench right now. [Our docs](https://docs.posit.co/resources/install-python.html) tell folks to install Python interpreters into `/opt/python/` but then says adding these to the `PATH` is optional. As of today, Positron will not find Python interpreters there unless they are added to the path.

**Comment by testlabauto** [comment]
### Verified Fixed
    Positron Version(s) : 2025.03.0-98
    OS Version          : OSX

### Test scenario(s) 

Looks great!  Include and exclude both functioned as expected for me.  Spoke with Sharon about the possibility of excluding all pyenvs except one and she indicated this scenario would be handled with an upcoming feature.

Link(s) to TestRail test cases run or created:
